### DISCUSSION TITLE
the difference between view and group

### DISCUSSION THREAD

**Discussion by armorhu** [skypjack/entt#638] [Q&A]
after i read the wiki（https://github.com/skypjack/entt/wiki/Crash-Course:-entity-component-system），i still feel unsure about when use view and when use group。I think they are the same。
So，what is the difference？

**Comment by skypjack** [comment]
Hi, sorry if I'm late. I wrote about groups [here](https://skypjack.github.io/2019-03-21-ecs-baf-part-2-insights/). They are just an optimization with its tradeoff.
The post should clear the fog. Let me know if you've still doubts after reading it. 👍

**Reply by armorhu** [reply]
thanks for u replay and  excellent tech article。I will read them carefully。But before that，I have anthor question and hope get some answer。
here it is：
Suppose i have 500 entities，and each of them have a component named MovmentComponent。
and i have a system to do the movement logic per frame。
auto view = register.view<MovementComponent>();
view.each([](MovementComponent& comp){
   //do the movement logic
})

but , as we know,not all the eneity need to be move because of the user's input. So we must add some flag to filter eneites
auto view = register.view<MovementComponent>();
view.each([](MovementComponent& comp){
   if(comp.needUpdateMovement){
   //do the movement logic
   }
})

So my question is ,is there better way to do this?  Is there something like tag for use to filter entites ? 
like:
auto view = register.view_withtagfilter<MovementComponent>("NeedUpdate");
view.each([](MovementComponent& comp){
   //do the movement logic
})

which is the best performance way?

**Comment by skypjack** [comment]
Well, you've components, use them. You don't want to iterate all `MovementComponent`s and check a flag. Instead, you want probably to iterate a `MovementComponent` and `DirtyFlagComponent` and only update the entities returned by this view. Most likely, they are going to be very low in count.

**Reply by armorhu** [reply]
II understand your idea。
But I think In over case，the number of dirty entities may be a lot in every game loop。
So，I am worried that the performance of dynamically adding or removing DirtyFlagComponent will be poor。
Do we have any test data compare Method A and Method B？
A：
view.each([](MovementComponent& comp){
if(comp.needUpdateMovement){
//do the movement logic
}
})
B：
//add or removing DirtyFlagComponent in some place.
group<MovementComponent& comp,DirtyFlagComponent>.each([](MovementComponent& comp,DirtyFlagComponent& dirtyComp){
    
   
})

**Reply by skypjack** [reply]
It depends on the underlying implementations. Every model has its pros and cons. In EnTT, adding or removing components is pretty cheap in general, but still. You can try and measure, of course.
In fact, you don't even have to remove, since you can just add, elaborate and clear the whole pool at once when you're done.

**Reply by armorhu** [reply]
I think I understand this problem enough。Thank you for your patient explanation。
on the other hand ， did you know anybody use entt in ue4 mobile project？
I haven't tried to upgrade the ios and android compilation of ue4 to c++17，I don't know what problem will be encountered ，
So if you could provide any information on this, I would be very grateful and save me a lot of time

**Reply by skypjack** [reply]
This may sound a little as a RTM advice, but the wiki contains an [_`EnTT` and `UE4`_ section](https://github.com/skypjack/entt/wiki/EnTT-and-Unreal-Engine). Probably that's a good place to start. 🙂 
Then I suggest to join the gitter channel or the discord server if you need help. 👍 Links on the README file.
